### ISSUE TITLE
Content-MD5 and request content validation

### ISSUE THREAD

**Issue by fxposter** [Asquera/warden-hmac-authentication#21]
In README you say that Content-MD5 header is always included in canonical representation (if it exists) but the library does not provide any validations that Content-MD5 is valid and Faraday middleware also does not add it if request body is not empty.

Are you planning to implement those features or if not, can you tell how would you implement them?

**Comment by gregory-m** [comment]
Also, how you assume to authorize POST requests. Current implementation don't check post body. So hacker can make request to same URL but with another body using same signature.

We have understood that you plan to use Content-MD5 header to avoid these situations.
Is that correct?

**Comment by Xylakant** [comment]
yes, you're right. Currently the library only includes content-md5 into the validation of the signature, but does not check that it actually matches the post body. That's currently up the application, but I guess you're right - it could at least be included in the faraday middleware, if not in the authenticator itself.

Let me think about that a little, I'll have to figure out how to do that. There might be some pitfalls such as 100-Continue headers etc. 

@gregory-m Authorizing POST-Requests can serve different purposes: 
- Authorize a client that does not own a secret to submit whatever data he wants and hand him a URL that's authorized for that purpose. In that case, validating the POST body is up to the application since you can't generate a URL that's tied to the body. This is fully supported right now. Uploads to Amazons S3 Service can be secured in a similar fashion.
- A client that knows his secret wants to post a specific piece of data. In that case, the client would include the content-md5 header. That's partially supported as described above: The header is checked for and included in the signature, but the check that the body matches the content-md5 header is not performed.
